### PR TITLE
Feature/watercannon puzzle

### PR DESCRIPTION
**Pull request by Lucas4185**

### CODE REVIEW

File: Assets/Scripts/Watercannon/RaycastWaterCannon.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        private void DrawLine()
+        {
+            var pointList = new List<Vector3>();
+            for (float ratio = 0; ratio <= 1; ratio += 1.0f / _vertexCount)
+            {
+                var tangentLineVertex1 = Vector3.Lerp(_point1.localPosition, _point2.localPosition, ratio);
+                var tangentLineVertex2 = Vector3.Lerp(_point2.localPosition, _point3.localPosition, ratio);
+                var bezierpoint = Vector3.Lerp(tangentLineVertex1, tangentLineVertex2, ratio);
+                pointList.Add(bezierpoint);
+            }
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/RaycastWaterCannon.cs:52]: In complex code please explain what and why you're doing

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        private CannonMovement _cm;
+
+        [SerializeField]
+        private SteamVR_Action_Boolean _trigger;
+        [SerializeField]
+        private SteamVR_Input_Sources _leftHand = SteamVR_Input_Sources.LeftHand;
+        [SerializeField]
+        private SteamVR_Input_Sources _rightHand = SteamVR_Input_Sources.RightHand;
+
+        private bool _includeChildren = true;
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/ShootWater.cs:32]: Isn't used?

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            _system.SetParticles(_particles, numParticlesAlive);
+        }
+
+        void InitializeIfNeeded()
+        {
+            if (_system == null)
+                _system = GetComponent<ParticleSystem>();
+
+            if (_particles == null || _particles.Length < _system.main.maxParticles)
+                _particles = new ParticleSystem.Particle[_system.main.maxParticles];
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/ShootWater.cs:84]: Weird bracket inconsistency?

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
+﻿using System.Collections;
+using System.Collections.Generic;
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/ShootWater.cs:2]: These two aren't used?

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            Debug.Log("Shooting water");
+            Debug.Log(_system.enableEmission);
+            if(_cm.puzzleStarted == true)
+            {
+                _system.enableEmission = true;
+            }
+            else
+            {
+                _system.enableEmission = false;
+            }
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/ShootWater.cs:62]: Consider inline statement?

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        private SteamVR_Input_Sources _leftHand = SteamVR_Input_Sources.LeftHand;
+        [SerializeField]
+        private SteamVR_Input_Sources _rightHand = SteamVR_Input_Sources.RightHand;
+
+        private bool _includeChildren = true;
+
+        private void Start()
+        {
+            _system.enableEmission = false;
+        }
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/ShootWater.cs:37]: Alex recently showed the => way of working, consider trying it!

File: Assets/Scripts/Watercannon/CannonMovement.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        [SerializeField]
+        private GameObject[] _hands;
+        [SerializeField]
+        private GameObject[] _handles;
+        [SerializeField]
+        private GameObject _waterCannon;
+        [SerializeField]
+        private GameObject _rotationTarget;
+
+        public bool puzzleStarted = false;
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/CannonMovement.cs:22]: Why isn't a SO used?

**Review by SjorsMaster — CHANGES_REQUESTED** [review]
Please look at the added comments and suggestions (Not all are mandatory), and fix the merge errors. (Did you make your branch up to date with the develop?)
image

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            else
+            {
+              _system.enableEmission = false;
+            }
+        }
+        //In this function the particle system wil be called to play and fire particles that with go down at a given speed
+        private void FiringWater()
+        {
+            Debug.Log("Shooting water");
+            Debug.Log(_system.enableEmission);
```
Comment (AlexJanson) [inline Assets/Scripts/Watercannon/ShootWater.cs:54]: Please remove the debug logs from the pr

File: Assets/Scripts/Watercannon/ShootWater.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+namespace LekAanDek.Puzzles.WaterCannon
+{
+    /// <summary>
+    /// This class checks if the player is pressing the trigger button and shoots water if he does
+    /// </summary>
+    [RequireComponent(typeof(ParticleSystem))]
+    public class ShootWater : MonoBehaviour
+    {
+        [SerializeField]
+        private BoolVariable startedWCPuzzle;
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/ShootWater.cs:14]: This does not comply with the naming convention, please add a prefix

File: Assets/Scripts/Watercannon/RaycastWaterCannon.cs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+        [SerializeField]
+        private BoolVariable startedWCPuzzle;
+
+        [SerializeField]
+        private Transform _point1;
+        [SerializeField]
+        private Transform _point2;
+        [SerializeField]
+        private Transform _point3;
```
Comment (SjorsMaster) [inline Assets/Scripts/Watercannon/RaycastWaterCannon.cs:21]: Please provide a tooltip as in the editor no one would know what these are for.

**Review by SjorsMaster — APPROVED** [review]
It's fine, approved

**Pull request merged**